compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp

```diff
@@ -78,240 +103,737 @@ static void hoistImmutableLoads(Region &region,
+  bool processLoopLikeOp(LoopLikeOpInterface loop);
+  // Process the given loop, hoisting redundant accessors. On success this
+  // returns the newly created loop and replaces the accessor map entry for
+  // the original loop with the new one.
+  Operation *hoistRedundantAccessors(LoopLikeOpInterface loop);
+  // Process all composite accessors (e.g. scf.for).
+  bool processCompositeAccessors();
+  // Process all of loads/stores for the given global.
+  bool processDirectAccessors(StringAttr globalName);
+  // Process all accessors by bucket in |bucketedAccessors|.
+  bool processGlobalAccessors();
+
+private:
+  // Move an operation right after the nearest blocking operation above |op|
+  // within its block. An operation is blocking if it *may* access the global
+  // referenced by |globalName|.
+  void moveOpUpInBlock(Operation *op, StringAttr globalName);
+  // Move an operation right before the nearest blocking operation below |op|
+  // within its block. An operation is blocking if it *may* access the global
+  // referenced by |globalName|.
+  void moveOpDownInBlock(Operation *op, StringAttr globalName);
+
+  StringAttr getGlobalName(Operation *op) const {
+    StringAttr globalName;
+    if (auto store = dyn_cast<IREE::Util::GlobalStoreOpInterface>(op)) {
+      globalName = store.getGlobalAttr().getAttr();
+    } else if (auto load = dyn_cast<IREE::Util::GlobalLoadOpInterface>(op)) {
+      globalName = load.getGlobalAttr().getAttr();
+    } else {
+      llvm_unreachable("Unhandled direct accessor type");
+    }
+    return globalName;
+  }
+
+  void decrementAccessorCount(StringAttr globalName, Operation *op,
+                              int64_t i = 1) {
+    if (accessorCounts.contains(op)) {
+      auto &map = accessorCounts[op];
+      DFX::SaturatedInteger newVal = map[globalName] - i;
+      LLVM_DEBUG({
+        llvm::dbgs() << "Decrementing accessor count of\n";
+        op->print(llvm::dbgs(), getAsmState());
+        llvm::dbgs() << "\n";
+        llvm::dbgs() << "for global " << globalName << " by " << i << " to "
+                     << newVal << "\n";
+      });
+      if (newVal == 0) {
+        map.erase(globalName);
+      } else {
+        map[globalName] = newVal;
+      }
+    }
+  }
+
+  void eraseAccessor(StringAttr globalName, Operation *op) {
+    Operation *parent = op->getParentOp();
+    // Decrement the accessor count for this global and propagate to all
+    // parents.
+    // TODO: This does not properly track accessor changes across function
+    // bounaries, however without hoisting across function boundaries, this
+    // has no effect on the current reached fixedpoint given that accessors
+    // are only ever removed one at a time from a pair within the same region.
+    while (parent) {
+      decrementAccessorCount(globalName, parent);
+      parent = parent->getParentOp();
+    }
+    op->erase();
+  }
+
+  bool operationDoesAccessGlobal(Operation *op, StringAttr globalName) const {
```

Comment:
Almost all of the non-analysis time is spent here, just querying whether or not an operation accesses a particular global. This is all cached analysis state so I'm not entirely sure how to make this much faster; maybe there are some better data structures than DenseMap. Otherwise, there might be a way to reorganize the processing of ops that saves some traversal of ops surrounding global loads/stores, but my attempts to improve this haven't worked at least.